### PR TITLE
DHCP labels for version 3.1.1h and above

### PR DESCRIPTION
**Pull request by shrsr**
Ability to add multiple DHCP policies

### CODE REVIEW

File: plugins/module_utils/mso.py [hunk 1]
```diff
@@ -907,6 +919,15 @@ def make_dhcp_label(self, data):
         ''' Create a DHCP policy from input '''
         if data is None:
             return None
+        if type(data) == list:
+            dhcps = []
+            for dhcp in data:
+                if 'dhcp_option_policy' in dhcp:
+                    dhcp_option_policy = dhcp['dhcp_option_policy']
```
Comment (anvitha-jain) [inline plugins/module_utils/mso.py:926]: can these two lines be put in one like
dhcp['dhcpOptionLabel'] = dhcp['dhcp_option_policy']
?

File: plugins/module_utils/mso.py [hunk 2]
```diff
@@ -879,7 +887,7 @@ def make_reference(self, data, reftype, schema_id, template):
             'templateName': template,
         }
 
-    def make_subnets(self, data, querier=True):
+    def make_subnets(self, data, querier=True, primary=True, virtual=True):
```
Comment (lhercot) [inline plugins/module_utils/mso.py:890]: I thought virtual and primary are not necessarily for all subnets in the list. This change is impact all current use of make_subnets.

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 3]
```diff
@@ -388,7 +411,8 @@
     - nm_add_bd_5.current.l2Stretch == true
     - nm_add_bd_5.current.l2UnknownUnicast == "flood"
     - nm_add_bd_5.current.l3MCast == false
-  when: version.current.version is version('3.1.1g', '<')
+  when: 
+  - version.current.version is version('3.1.1g', '<')
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:415]: If single condition, they should be on same line as the when. See below for the others that are similar too.

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 4]
```diff
@@ -883,6 +912,277 @@
...
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy3
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option2
+          version: 1
+    state: present
+  register: nm_bd_dhcp_policies
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:961]: Shouldn't this have a WHEN condition?

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 5]
```diff
@@ -883,6 +912,277 @@
...
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy2
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+    state: present
+  register: change_nm_bd_dhcp_policies
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1043]: Shouldn't this have a WHEN condition?

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 6]
```diff
@@ -883,6 +912,277 @@
...
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy2
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+    state: present
+  register: nm_bd_new_subnet_options
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1105]: Shouldn't this have a WHEN condition?

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 7]
```diff
@@ -883,6 +912,277 @@
...
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy2
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+    state: present
+  register: nm_bd_subnet_second_primary
+  ignore_errors: yes
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1178]: Shouldn't this have a WHEN condition?

Comment (lhercot) [reply tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1178]: If it is a successive list of tasks that need the same condition you can also use a block statement like: https://github.com/CiscoDevNet/ansible-mso/blob/0ce486646619105c3eb2cbe85762ee72f86fb7a3/tests/integration/targets/mso_user/tasks/main.yml#L30

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 8]
```diff
@@ -1009,6 +1309,131 @@
...
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy2
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+    state: query
+  register: query_nm_bd_dhcp_policies
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1353]: Shouldn't this have a WHEN condition?

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml [hunk 9]
```diff
@@ -1009,6 +1309,131 @@
...
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+      - name: ansible_test_dhcp_policy2
+        version: 1
+        dhcp_option_policy:
+          name: ansible_test_dhcp_policy_option1
+          version: 1
+    state: query
+  register: query_bd_new_subnet_options
```
Comment (lhercot) [inline tests/integration/targets/mso_schema_template_bd/tasks/main.yml:1415]: Shouldn't this have a WHEN condition?

File: plugins/module_utils/mso.py (hunk 2 above)
Comment (lhercot) [reply plugins/module_utils/mso.py:890]: For example are those available under the subnet inside an EPG?

**Review by lhercot — CHANGES_REQUESTED** [review]

File: plugins/modules/mso_schema_template_bd_dhcp_policy.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        version=dict(type='int'),
+        dhcp_option_policy=dict(type='dict', options=mso_dhcp_option_spec()),
+        state=dict(type='str', default='present', choices=['absent', 'present', 'query']),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ['state', 'absent', ['dhcp_policy', 'version']],
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_bd_dhcp_policy.py:153]: Why do you need version for absent?

File: tests/integration/targets/mso_schema_template_bd/tasks/main.yml (hunk 3 above)
Comment (shrsr) [reply tests/integration/targets/mso_schema_template_bd/tasks/main.yml:415]: ok

File: plugins/module_utils/mso.py (hunk 2 above)
Comment (shrsr) [reply plugins/module_utils/mso.py]: Added is_bd_subnet option to cover all 3 parameters.

File: plugins/module_utils/mso.py (hunk 1 above)
Comment (shrsr) [reply plugins/module_utils/mso.py:922]: we could

Comment (lhercot) [reply plugins/module_utils/mso.py:922]: dhcp['dhcpOptionLabel'] = dhcp.get('dhcp_option_policy')

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**